### PR TITLE
Adding `AutomaticSpeechRecognitionPipeline`.

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

- Because we added everything to enable this pipeline, we probably
should add it to `transformers`.
- This PR tries to limit the scope and focuses only on the pipeline part
(what should go in, and out).
- The tests are very specific for S2T and Wav2vec2 to make sure both
architectures are supported by the pipeline. We don't use the mixin for
tests right now, because that requires more work in the `pipeline`
function (will be done in a follow up PR).
- Unsure about the "helper" function `ffmpeg_read`. It makes a lot of
  sense from a user perspective, it does not add any additional
dependencies (as in hard dependency, because users can always use their
own load mechanism). Meanwhile, it feels slightly clunky to have so much
optional preprocessing.
- The pipeline is not done to support streaming audio right now.

# Future work:

- Add `automatic-speech-recognition` as a `task`. And add the
FeatureExtractor.from_pretrained within `pipeline` function.
- Add small models within tests
- Add the Mixin to tests.
- Make the logic between ForCTC vs ForConditionalGeneration better.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@LysandreJik
@patrickvonplaten

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/pipelines/automatic_speech_recognition.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+from .base import Pipeline
+
+
+if TYPE_CHECKING:
+    from ..models.auto import AutoFeatureExtractor
+
+logger = logging.get_logger(__name__)
+
+
+def ffmpeg_read(bpayload: bytes, sampling_rate: int) -> np.array:
```
Comment (LysandreJik) [inline src/transformers/pipelines/automatic_speech_recognition.py:16]: Not too sure about this function, but if we enable it we should have it gracefully fail if the dependency is not found on the system

File: src/transformers/pipelines/automatic_speech_recognition.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        """
+        Classify the sequence(s) given as inputs. See the :obj:`~transformers.AutomaticSpeechRecognitionPipeline`
+        documentation for more information.
+
+        Args:
+            inputs (:obj:`np.ndarray` or :obj:`bytes` or :obj:`str`):
+                The inputs is either a raw waveform (:obj:`np.ndarray`) at the correct sampling rate (no further check will be done) or a :obj:`str` that is
+                the filename of the audio file, the file will be read at the correct sampling rate to get the waveform using `ffmpeg`.
+                This requires `ffmpeg` to be installed on the system. If `inputs` is :obj:`bytes` it is supposed to be the content of an audio file and is
+                interpreted by `ffmpeg` in the same way.
```
Comment (LysandreJik) [inline src/transformers/pipelines/automatic_speech_recognition.py:99]: The pipeline doesn't handle batching - could the expected dimensions of this `inputs` variable when passed in as an `np.ndarray` be explicit?

File: tests/test_pipelines_automatic_speech_recognition.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
+# Copyright 2020 The HuggingFace Team. All rights reserved.
```
Comment (LysandreJik) [inline tests/test_pipelines_automatic_speech_recognition.py:1]:
```suggestion
# Copyright 2021 The HuggingFace Team. All rights reserved.
```

File: src/transformers/pipelines/automatic_speech_recognition.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+class AutomaticSpeechRecognitionPipeline(Pipeline):
+    """
+    Pipeline that aims at extracting spoken text contained within some audio.
+
+    The input can be either a raw waveform or a audio file. In case of the audio file, ffmpeg should
+    be installed for the read function requires it (to support many formats).
+    """
+
+    def __init__(self, feature_extractor: "AutoFeatureExtractor", *args, **kwargs):
```
Comment (LysandreJik) [inline src/transformers/pipelines/automatic_speech_recognition.py:57]:
```suggestion
    def __init__(self, feature_extractor: "SequenceFeatureExtractor", *args, **kwargs):
```

File: src/transformers/pipelines/automatic_speech_recognition.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+                The feature extractor that will be used by the pipeline to encode waveform for the model.
+            modelcard (:obj:`str` or :class:`~transformers.ModelCard`, `optional`):
+                Model card attributed to the model for this pipeline.
+            framework (:obj:`str`, `optional`):
+                The framework to use, either :obj:`"pt"` for PyTorch or :obj:`"tf"` for TensorFlow. The specified framework
+                must be installed.
+
+                If no framework is specified, will default to the one currently installed. If no framework is specified and
+                both frameworks are installed, will default to the framework of the :obj:`model`, or to PyTorch if no model
+                is provided.
```
Comment (LysandreJik) [inline src/transformers/pipelines/automatic_speech_recognition.py:77]: It should probably just fail if something other than `pt` is passed in, as some very specific PyTorch instructions are set in the call methods, and the models don't have a TensorFlow counterpart.

See an example of this in the `TableQuestionAnsweringPipeline`:
https://github.com/huggingface/transformers/blob/cfd2eaa8cf82da8581825c6592b66d2789c5bc53/src/transformers/pipelines/table_question_answering.py#L86-L87

File: src/transformers/pipelines/automatic_speech_recognition.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
+import subprocess
```
Comment (LysandreJik) [inline src/transformers/pipelines/automatic_speech_recognition.py:1]: Should have the copyright above

**Review by LysandreJik — COMMENTED** [review]
Thanks for working on it!

It is very specific to S2T and Wav2Vec2 but I don't think that's too much of an issue, we can adapt later.

Could you add this pipeline to:
- the documentation
- the main init
- the `pipeline` factory method

We will also probably need a new auto model.

File: src/transformers/pipelines/automatic_speech_recognition.py (hunk 1 above)
Comment (Narsil) [reply src/transformers/pipelines/automatic_speech_recognition.py:16]: I'm not sure either. 

We already seem to use `soundfile` but `soundfile` won't read as many files + will fail if the sampling rate is incorrect. It feels pretty bad UX for pipeline users.

File: src/transformers/pipelines/automatic_speech_recognition.py (hunk 2 above)
Comment (Narsil) [reply src/transformers/pipelines/automatic_speech_recognition.py:99]: Well I think pipelines should try to avoid doing real batching anyway : 
https://github.com/huggingface/transformers/pull/11251#pullrequestreview-637488364

I could add the `list` behavior that exists ìn other pipelines, but I wouldn't implement it as batching (here padding will be much worse than with text).

I really full funny implementing :
```python
if not isinstance(xX, list): 
    inputs = [inputs]
.... 
for input_ in inputs: 
    results.append(do_something(input_))

if len(results) == 1:
   results = results[0]
```
everywhere. It's error-prone and confusing to me. For instance it will return an object and not a list if I passe a single item list as input, where I would expect to receive a single list output.

Because this pipeline does not currently exist, I was hesitant to put this directly here. There is no backward compatibility yet, I wanted to avoid introducing something which feels like a bad pattern.
Coherence is also important, so I can definitely add it if coherence trumps correctness in your opinion.

File: src/transformers/pipelines/automatic_speech_recognition.py (hunk 5 above)
Comment (Narsil) [reply src/transformers/pipelines/automatic_speech_recognition.py:77]: Good point !

**Comment by Narsil** [comment]
> Thanks for working on it!
> 
> It is very specific to S2T and Wav2Vec2 but I don't think that's too much of an issue, we can adapt later.
> 
> Could you add this pipeline to:
> 
>     * the documentation

Yes !
> 
>     * the main init

Yes

> 
>     * the `pipeline` factory method

I wanted to defer this into a follow-up PR, because of the AutoModel quirckness and to avoid making super long PRs.
If you feel we can't have it as a separate PR, I'll start including the work directly here.
> 
> 
> We will also probably need a new auto model.

File: src/transformers/pipelines/automatic_speech_recognition.py (hunk 6 above)
Comment (Narsil) [reply src/transformers/pipelines/automatic_speech_recognition.py:1]: Ok

File: src/transformers/pipelines/automatic_speech_recognition.py (hunk 4 above)
Comment (Narsil) [reply src/transformers/pipelines/automatic_speech_recognition.py:57]: Thanks

File: src/transformers/pipelines/automatic_speech_recognition.py (hunk 2 above)
Comment (LysandreJik) [reply src/transformers/pipelines/automatic_speech_recognition.py:118]: I definitely understand your comment regarding batching, I wasn't proposing enabling batching on the pipeline - just specifying what are the expected dimensions of `inputs`!

Comment (Narsil) [reply src/transformers/pipelines/automatic_speech_recognition.py]: You mean adding so asserts on dimensions ?

Comment (Narsil) [reply src/transformers/pipelines/automatic_speech_recognition.py]: Just the docs I guess :)

File: src/transformers/pipelines/automatic_speech_recognition.py (hunk 1 above)
Comment (anton-l) [reply src/transformers/pipelines/automatic_speech_recognition.py:29]: How about `librosa.load(path, sr=sampling_rate, mono=True)`? https://librosa.org/doc/latest/generated/librosa.load.html#librosa.load
It uses `soundfile` as a default backend and falls back on `audioread`+`ffmpeg` for unsupported codecs like mp3.
Also, loading and resampling+mono in a single line is pretty nice :)

Comment (Narsil) [reply src/transformers/pipelines/automatic_speech_recognition.py:29]: `librosa` is not a dependency currently, I'm more than hesitant to add it for just that feature.

Also `librosa` api enforces to use a filename which is pretty bad in the context of an API (where hitting disk is never desirable). It does fallback to ffmpeg anyway, in pretty much what `ffmpeg_read` does. (Actually if you use soundfile and resample later then you're also wasting a lot of time, it's much better if the file can be read at the correct sampling rate directly).

The main issue I think is : should we have such functions in `transformers` ? 
It makes life easier on users which is Priority number 1 of the lib (especially pipelines).
At the same time, I don't think we want to reinvent all image/audio preprocessing.

Note: `torchaudio` is another alternative to `librosa` to consider too. (it doesn't do automatic resampling afaik)

Comment (julien-c) [reply src/transformers/pipelines/automatic_speech_recognition.py:29]: @mpariente and @jonashaag might also want to chime in on this because I remember having discussions on this subject with them

Comment (jonashaag) [reply src/transformers/pipelines/automatic_speech_recognition.py:29]: The question is what is your priority.

- If you are looking to minimise dependencies while being able to read a lot of file types, this way is probably best. (Could switch the manual ffmpeg invocation for the `audioread` library.)
- If you are looking to minimise load time, then trying soundfile + [julius](https://github.com/adefossez/julius) is probably the best choice. (Could also add a fallback to ffmpeg.)
- To keep things as simple as possible and to write is little code as possible, just use `librosa.load` which tries soundfile first and then falls back to ffmpeg.

> (Actually if you use soundfile and resample later then you're also wasting a lot of time, it's much better if the file can be read at the correct sampling rate directly).

I don't think this is correct since ffmpeg also needs to do resampling, so it depends on the speed of ffmpeg resampling implementation vs. your resampling implementation.

Comment (patrickvonplaten) [reply src/transformers/pipelines/automatic_speech_recognition.py:29]: We should build a library for reading audio files in Rust :-)

File: src/transformers/pipelines/automatic_speech_recognition.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        raise ValueError("Malformed soundfile")
+    return audio
+
+
+class AutomaticSpeechRecognitionPipeline(Pipeline):
+    """
+    Pipeline that aims at extracting spoken text contained within some audio.
+
+    The input can be either a raw waveform or a audio file. In case of the audio file, ffmpeg should be installed for
+    the read function requires it (to support many formats).
```
Comment (patrickvonplaten) [inline src/transformers/pipelines/automatic_speech_recognition.py:70]:
```suggestion
    to support multiple audio formats
```

File: src/transformers/pipelines/automatic_speech_recognition.py [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+        """
+        Arguments:
+            model (:obj:`~transformers.PreTrainedModel` or :obj:`~transformers.TFPreTrainedModel`):
+                The model that will be used by the pipeline to make predictions. This needs to be a model inheriting
+                from :class:`~transformers.PreTrainedModel` for PyTorch and :class:`~transformers.TFPreTrainedModel`
+                for TensorFlow.
+            tokenizer (:obj:`~transformers.PreTrainedTokenizer`):
+                The tokenizer that will be used by the pipeline to encode data for the model. This object inherits from
+                :class:`~transformers.PreTrainedTokenizer`.
+            feature_extractor (:obj:`~transformers.SequenceFeatureExtractor`):
```
Comment (patrickvonplaten) [inline src/transformers/pipelines/automatic_speech_recognition.py:83]: maybe put this one first since it's also first in the function signature?

File: src/transformers/pipelines/automatic_speech_recognition.py [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+                the filename of the audio file, the file will be read at the correct sampling rate to get the waveform
+                using `ffmpeg`. This requires `ffmpeg` to be installed on the system. If `inputs` is :obj:`bytes` it is
+                supposed to be the content of an audio file and is interpreted by `ffmpeg` in the same way.
+
+        Return:
+            A :obj:`dict` with the following keys:
+
+            - **text** (:obj:`str`) -- The recognized text.
+        """
+        if isinstance(inputs, str):
```
Comment (patrickvonplaten) [inline src/transformers/pipelines/automatic_speech_recognition.py:126]: I would also add an assert here that `inputs` cannot be batched

File: tests/test_pipelines_automatic_speech_recognition.py [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+from transformers import AutoFeatureExtractor, AutoTokenizer, Speech2TextForConditionalGeneration, Wav2Vec2ForCTC
+from transformers.pipelines import AutomaticSpeechRecognitionPipeline
+from transformers.testing_utils import require_torch, require_torchaudio, slow
+
+
+# from .test_pipelines_common import CustomInputPipelineCommonMixin
+
+
+class AutomaticSpeechRecognitionPipelineTests(unittest.TestCase):
```
Comment (patrickvonplaten) [inline tests/test_pipelines_automatic_speech_recognition.py:25]: we should add a `require_datasets` here I think

**Review by patrickvonplaten — APPROVED** [review]
Thanks a lot for adding this!

File: src/transformers/pipelines/automatic_speech_recognition.py (hunk 8 above)
Comment (Narsil) [reply src/transformers/pipelines/automatic_speech_recognition.py:83]: Ok.

File: src/transformers/pipelines/automatic_speech_recognition.py (hunk 9 above)
Comment (Narsil) [reply src/transformers/pipelines/automatic_speech_recognition.py:126]: I´ve added those, it´s right after the attempt to load from filename or from bytes.

File: tests/test_pipelines_automatic_speech_recognition.py (hunk 10 above)
Comment (Narsil) [reply tests/test_pipelines_automatic_speech_recognition.py:25]: Done !

File: src/transformers/pipelines/automatic_speech_recognition.py (hunk 1 above)
Comment (julien-c) [reply src/transformers/pipelines/automatic_speech_recognition.py:29]: > We should build a library for reading audio files in Rust with Python bindings

Famous last words

Comment (julien-c) [reply src/transformers/pipelines/automatic_speech_recognition.py:29]: Thanks so much @jonashaag! Am I wrong in remembering that `julius`' resampling was more or less merged into recent versions of torchaudio? (also cc'ing @adefossez)

Comment (mpariente) [reply src/transformers/pipelines/automatic_speech_recognition.py:29]: Yes it was!

Comment (Narsil) [reply src/transformers/pipelines/automatic_speech_recognition.py:29]: @LysandreJik do you have any opinions ? 

I'll vote in favor of this current implementation, it's pretty small code-wise, is probably the largest audio support of the proposed alternatives.

Comment (LysandreJik) [reply src/transformers/pipelines/automatic_speech_recognition.py:29]: Ok for me, let's go with the current proposition.

**Review by LysandreJik — APPROVED** [review]

File: src/transformers/pipelines/automatic_speech_recognition.py (hunk 1 above)
Comment (adefossez) [reply src/transformers/pipelines/automatic_speech_recognition.py:29]: I'm joining a bit late the conversation now that this is merged. ffmpeg is great but a bit of a heavy dependency and also it is not possible to ensure which format it will support, depending on how it was installed.
Also, ffmpeg cannot be installed automatically as a dependency through pip, or requirements in a `setup.py` file. 

For the latest version of Demucs, I ended up doing a combination of torchaudio (doing resampling with `julius` and manual channel conversion) + fallback to ffmpeg if torchaudio failed, because torchaudio on Windows has a really limited support (no mp3, although that will change soon), so that on Linux and Mac it will work with a simple `pip install`, while on Windows, I still need people to use Anaconda to install ffmpeg.

https://github.com/facebookresearch/demucs/blob/master/demucs/separate.py#L21 (in there it first tried torchaudio, then fallback to `AudioFile` which is a similar wrapper around ffmpeg).

Comment (julien-c) [reply src/transformers/pipelines/automatic_speech_recognition.py:29]: thanks @adefossez
